doc/util/collect-syntax.py

```diff
@@ -0,0 +1,327 @@
+  out.write('.. code-block:: syntax\n')
+  out.write('\n')
+  global _productions
+  for ident in sorted(_productions.keys()):
+    _, _, blob = _productions[ident]
+    out.write(blob)
+    out.write('\n')
+    pass
+  return
+
+
+def is_meta_keyword(ident):
+  return ident == 'OPT' or ident == '-' or ident == '_' or ident == 'b'
+
+
+def is_meta_sentence(line):
+  return line.lstrip().startswith('any character except')
+
+
+def get_nonterminals(line):
+  # Regex trick, third capture group contains valid nonterminals.
+  regex_groups = re.findall(_match_nonterminal, line)
+  result = []
+  for group in regex_groups:
+    quoted, escaped, ok = group
+    # ''' uninterpreted-single-quote-delimited-characters '''
+    # This odd nonterminal appears in 'uninterpreted-string-literal' and is
+    # triple quoted. Instead of handling it with a fourth regex group in
+    # _match_nonterminal, special case it here.
+    if quoted:
+      clean = quoted.lstrip('\' ').rstrip('\' ')
+      if clean == 'uninterpreted-single-quote-delimited-characters':
+        result.append(clean)
+        continue
+    if ok:
+      result.append(ok)
+  return result
+
+  
+def get_nonterminals_in_expansion(ident, stack, applied):
+  global _productions
+  fname, lineno, blob = _productions[ident]
+  vprint('Inspecting expansions of:', ident)
+  vprint(blob)
+  nonterminals = []
+  for line in blob.split('\n'):
+    if re.match(_match_production, line) or is_meta_sentence(line):
+      continue
+    nonterminals += get_nonterminals(line)
+  vprint('Found', len(nonterminals), 'nonterminals')
+  for n in nonterminals:
+    if is_meta_keyword(n):
+      continue
+    if not n in _productions:
+      msg = 'Nonterminal not in production list: ' + str(n)
+      vprint(msg)
+      raise Exception(msg)
+    if not n in applied:
+      stack.append(n)
+      applied.add(n)
+  return
+
+
+def output_productions_application_order(out):
+  out.write('.. _Productions_Listed_Application_Order:\n')
+  out.write('\n')
+  title = 'Productions Listed in Application Order'
+  out.write(title + '\n')
+  out.write((len(title) * '-') + '\n')
+  out.write('\n')
+  out.write('.. code-block:: syntax\n')
+  out.write('\n')
+  global _productions
+  start = 'module-declaration-statement'
+  vprint('Computing application-first ordering starting with:', start)
+  stack = [start]
+  applied = {start}
+  # Do a breadth-first search over all productions...
+  while stack:
+    ident = stack.pop(0)
+    get_nonterminals_in_expansion(ident, stack, applied)
+    _, _, blob = _productions[ident]
+    out.write(blob)
+    out.write('\n')
+  # Consider unapplied productions to be a fatal error...
+  unapplied_count = len(_productions) - len(applied)
+  if unapplied_count != 0:
+    msg = str(unapplied_count) + ' nonterminals are never applied'
+    vprint(msg)
+    for ident in _productions.keys():
+      if not ident in applied:
+        vprint(ident)
+    raise Exception(msg)
+  return
+
+
+def main(args):
```

Comment:
You should not need to use the venv/chpldeps to run this script but we are using a convention there that `main` has no arguments (for things that are run through chpldeps). You can rename this one as `main_arguments`  and create `main` that is the body of the `if __name__ == '__main__'`. While it does not matter for this script I think it would be nice to follow the convention here.